### PR TITLE
Add features for v2.11.1

### PR DESCRIPTION
**Pull request by schlunma**
## Description

This PR adds all feature for the bugfix release v2.11.1 which enables lazy computations in more preprocessors and allows installing the latests version of various dependencies, including Iris ([v3.11.0](https://github.com/SciTools/iris/releases/tag/v3.11.0)).

Changelog and update of version number will be included in an upcoming PR.

This has been tested with various real recipes, all output looks good.

***

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [x] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [x] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [x] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [x] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] Any changed [dependencies have been added or removed](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#dependencies) correctly
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [x] [🛠][1] ~All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful~ Successful GitHub actions runs: [here](https://github.com/ESMValGroup/ESMValCore/actions/runs/12009810759/job/33475328604?pr=2585) and [here](https://github.com/ESMValGroup/ESMValCore/actions/runs/12009810757/job/33475328633?pr=2585).

Expected test fails:
- Pre-commit: has not been set up for v2.11.1
- CircleCI tests: this uses a docker container based on the ESMValCore main branch to run the tests, which includes numpy > 2. During the test, pip tries to install numpy < 2, which then [fails](https://app.circleci.com/pipelines/github/ESMValGroup/ESMValCore/12114/workflows/9b435a0f-da7a-4c26-a70e-dacda4c3a2ac/jobs/50812?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link&utm_content=summary).
- Codacy fails, but shows [no error](https://app.codacy.com/gh/ESMValGroup/ESMValCore/pull-requests/2585/issues). I think this is due to the missing coverage report (cause by the failing CircleCI tests).

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

**Comment by schlunma** [comment]
All my tests were successful! GitHub action tests are also green:

- https://github.com/ESMValGroup/ESMValCore/actions/runs/12009810759/job/33475328604?pr=2585
- https://github.com/ESMValGroup/ESMValCore/actions/runs/12009810757/job/33475328633?pr=2585

All other test fails are expected (see PR description). This is good to go ✅

**Review by valeriupredoi — CHANGES_REQUESTED** [review]
Thanks, Manu! No particular (point-wise changes required), but two main things I'd like to clarify before going forward:
- the pin to pandas is now removed bc you folded in the code changes that we made for the use of the latest pandas (sorry, I completely forgot what were those)
- v2.11.0 **was not** released with support for Python 2.12, but this bugfix release will be, in principle, I am OK with that, but for that we need to set a few more pins in the env file (they work without them due to mamba's penchant for latest versions, but we ought to set the pins to be 100% sure we don't end up in a pickle, at some point) - do you want to go ahead with this, or you remove 3.12 and pin python<3.12 as in the release?
:beer:

**Comment by schlunma** [comment]
>     * the pin to pandas is now removed bc you folded in the code changes that we made for the use of the latest pandas (sorry, I completely forgot what were those)

Yes, https://github.com/ESMValGroup/ESMValCore/pulls?q=is%3Apr+pandas+is%3Aclosed is included!
 
>     * v2.11.0 **was not** released with support for Python 2.12, but this bugfix release will be, in principle, I am OK with that, but for that we need to set a few more pins in the env file (they work without them due to mamba's penchant for latest versions, but we ought to set the pins to be 100% sure we don't end up in a pickle, at some point) - do you want to go ahead with this, or you remove 3.12 and pin python<3.12 as in the release?

Good point! I think it's nice to support Python 3.12, even though in practice this won't get installed since ESMValTool v2.11 does not support it. What kinds of pins do we need?

**Comment by valeriupredoi** [comment]
actually, having checked what we need for Python 3.12 from the initial PR https://github.com/ESMValGroup/ESMValCore/pull/2228/files , there is very little we should do here, and that may impact functionality: just remove `compilers` - nothing else that jumps :grin:

**Comment by valeriupredoi** [comment]
> Yes, https://github.com/ESMValGroup/ESMValCore/pulls?q=is%3Apr+pandas+is%3Aclosed is included!

Perfect, cheers :beer:

File: environment.yml
Comment (valeriupredoi) [inline environment.yml:11]:
```suggestion
```

**Review by valeriupredoi — APPROVED** [review]
very nice, Manu! Think this is good to go - I'll arrange a releae for you off at the feedstock after we merge it

**Comment by schlunma** [comment]
Great, thanks!! Will merge once the tests finished, and then open a PR for the changelog ✅

**Comment by valeriupredoi** [comment]
good call, forgot about the Changelog :laughing:

**Pull request merged**
